File: src/main/java/answer/chapter5/Problem5_4.java
package answer.chapter5;

import jm.app.algebra.Matrix;
import jm.app.optimize.SteepestDescentOptimizer;
import jm.app.optimize.TargetFunction;

import java.math.BigDecimal;

/**
 * 利用SteepestDescentOpitmizer编写程序求出5x1-2x2+3x3-6x4=0的一个可行解。
 */
public class Problem5_4 {

    static class MyTargetFunction implements TargetFunction {
        @Override
        public BigDecimal fun(Matrix params, Matrix args) {
            double x1 = params.getValue(0, 0).doubleValue();
            double x2 = params.getValue(0, 1).doubleValue();
            double x3 = params.getValue(0, 2).doubleValue();
            double x4 = params.getValue(0, 3).doubleValue();
            double a1 = args.getValue(0, 0).doubleValue();
            double a2 = args.getValue(0, 1).doubleValue();
            double a3 = args.getValue(0, 2).doubleValue();
            double a4 = args.getValue(0, 3).doubleValue();
            double result = Math.pow(a1 * x1 + a2 * x2 + a3 * x3 + a4 * x4, 2);
            return new BigDecimal(result);
        }
    }

    public static void main(String[] args) {
        Matrix arguments = new Matrix(1, 4);
        arguments.setValue(0, 0, 5);
        arguments.setValue(0, 1, -2);
        arguments.setValue(0, 2, 3);
        arguments.setValue(0, 3, -6);
        Matrix params = new Matrix(1, 4);
        params.setValue(0, 0, 0.3);
        params.setValue(0, 1, 0.3);
        params.setValue(0, 2, 0.5);
        params.setValue(0, 3, 0.2);
        MyTargetFunction targetFunction = new MyTargetFunction();
        SteepestDescentOptimizer steepestDescentOptimizer = new SteepestDescentOptimizer();
        steepestDescentOptimizer.setLearningRate(new BigDecimal(0.0001));
        Matrix optimizedParams = steepestDescentOptimizer.optimize(targetFunction, params, arguments);
        System.out.println(optimizedParams);

    }

}

File: py/answer/chapter5/problem5_4.py
import py.optimize.target_function as tf
import py.optimize.steepest_descent_optimizer as sdo
import numpy as np


class TestTargetFunction(tf.TargetFunction):

    def fun(self, params, args):
        x1 = params[0]
        x2 = params[1]
        x3 = params[2]
        x4 = params[3]
        a1 = args[0]
        a2 = args[1]
        a3 = args[2]
        a4 = args[3]
        res = np.power(a1 * x1 + a2 * x2 + a3 * x3 + a4 * x4, 2)
        return res

def test_optimizer():
    params = np.array([0.3, 0.3, 0.5, 0.2])
    args = np.array([5, -2, 3, -6])
    target_function = TestTargetFunction()
    optimizer = sdo.SteepestDescentOptimizer(epoch_num= 10000)
    optimized_params = optimizer.optimize_with_args(target_function, params, args)
    print(optimized_params)

if __name__ == "__main__":
    test_optimizer()